graphql_compiler/ast_manipulation.py

```diff
@@ -96,22 +119,68 @@ def get_only_selection_from_ast(ast, desired_error_type):
     return selections[0]
 
 
-def get_ast_with_non_null_stripped(ast):
+def get_ast_with_non_null_stripped(ast: TypeNode) -> Union[ListTypeNode, NamedTypeNode]:
     """Strip a NonNullType layer around the AST if there is one, return the underlying AST."""
+    result: TypeNode
+
     if isinstance(ast, NonNullTypeNode):
         stripped_ast = ast.type
         if isinstance(stripped_ast, NonNullTypeNode):
             raise AssertionError(
                 "NonNullType is unexpectedly found to wrap around another NonNullType in AST "
                 "{}, which is not allowed.".format(ast)
             )
-        return stripped_ast
+        result = stripped_ast
     else:
-        return ast
+        result = ast
+
+    if not isinstance(result, (ListTypeNode, NamedTypeNode)):
+        raise AssertionError(
+            f"Expected the result to be either a ListTypeNode or a NamedTypeNode, but instead "
+            f"found: {result}"
+        )
 
+    return result
 
-def get_ast_with_non_null_and_list_stripped(ast):
+
+def get_ast_with_non_null_and_list_stripped(ast: TypeNode) -> NamedTypeNode:
     """Strip any NonNullType or List layers around the AST, return the underlying AST."""
     while isinstance(ast, (NonNullTypeNode, ListTypeNode)):
         ast = ast.type
+
+    if not isinstance(ast, NamedTypeNode):
+        raise AssertionError(
+            f"Expected the AST value to be a NamedTypeNode, but unexpectedly instead found: {ast}"
+        )
+
+    return ast
+
+
+def assert_selection_is_a_field_node(ast: Any) -> FieldNode:
```

Comment:
I like the idea — perhaps `checked_cast` as an alternative name, to make the answer to "how is this safer" a bit more explicit?

Reply:
`checked_cast()` doesn't work with `Union` because `Union` is not a `Type`, so `Type[T]` doesn't match.

For bounded-length tuples, this is achievable:
```python
def checked_cast_to_union2(target_types: Tuple[Type[T_A], Type[T_B]], value: Any) -> Union[T_A, T_B]:
    if not isinstance(value, target_types):
        raise AssertionError()

    return value
```

I don't know of a way to make `target_types: Tuple[Type, ...]` work and connect the types it contains to the types inside the `Union` result.

Given that we're somewhat unlikely to have unions of more than like 3 or so things, I'm fine with adding `checked_cast_to_union2()` and in the future potentially adding a `checked_cast_to_union3()` function that takes a 3-tuple as necessary.

What do you think?